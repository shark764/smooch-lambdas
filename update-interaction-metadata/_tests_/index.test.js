const axios = require('axios');
const uuidv1 = require('uuid/v1');

jest.mock('axios');
jest.mock('uuid/v1');

global.process.env = {
  AWS_REGION: 'us-east-1',
  ENVIRONMENT: 'dev',
  DOMAIN: 'domain',
  smooch_api_url: 'mock-smooch-api-url',
};
uuidv1.mockImplementation(() => '7534c040-534d-11ea-8aa0-c32d6a748e46');

const event = {
  Records: [{
    body: JSON.stringify({
      tenantId: '250faddb-9723-403a-9bd5-3ca710cb26e5',
      interactionId: '667802d8-2260-436c-958a-2ee0f71f73f0',
      source: 'source',
      metadata: 'metadata',
    }),
  }],
};

const mockGetSecretValue = jest.fn(() => {})
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        username: 'UserName',
        password: 'Password',
      }),
    }),
  }));

axios.mockImplementation(() => ({
  data: {},
}));

jest.mock('aws-sdk', () => ({
  SecretsManager: jest.fn().mockImplementation(() => ({
    getSecretValue: mockGetSecretValue,
  })),
}));

const { handler } = require('../index');

describe('update-interaction-metadata', () => {
  describe('Everything is successful', () => {
    it('when the code runs without any error', async () => {
      const result = await handler(event);
      expect(result).toBeUndefined();
    });

    describe('Walkthrough', () => {
      beforeAll(async () => {
        jest.clearAllMocks();
        await handler(event);
      });
      it('passes in the correct arguments to secretsClient.getSecretValue() to get cx credentials', async () => {
        expect(mockGetSecretValue.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to axios', async () => {
        expect(axios.mock.calls).toMatchSnapshot();
      });
    });
  });
  it('throws an error when there is a problem retrieving cx credentials', async () => {
    try {
      mockGetSecretValue.mockRejectedValueOnce(new Error());
      await handler(event);
    } catch (error) {
      expect(Promise.reject(new Error('Error retrieving cx credentials'))).rejects.toThrowErrorMatchingSnapshot();
    }
  });

  it('throws an error when there is a problem updating interaction metadata', async () => {
    try {
      axios.mockRejectedValueOnce(new Error());
      await handler(event);
    } catch (error) {
      expect(Promise.reject(new Error('Failed to update interaction metadata'))).rejects.toThrowErrorMatchingSnapshot();
    }
  });
});
