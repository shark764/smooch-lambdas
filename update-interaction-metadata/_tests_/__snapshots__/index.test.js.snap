// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update-interaction-metadata Everything is successful Walkthrough passes in the correct arguments to axios 1`] = `
Array [
  Array [
    Object {
      "auth": Object {
        "password": "Password",
        "username": "UserName",
      },
      "data": Object {
        "metadata": "metadata",
        "source": "source",
      },
      "method": "post",
      "url": "https://us-east-1-dev-edge.domain/v1/tenants/250faddb-9723-403a-9bd5-3ca710cb26e5/interactions/667802d8-2260-436c-958a-2ee0f71f73f0/metadata?id=7534c040-534d-11ea-8aa0-c32d6a748e46",
    },
  ],
]
`;

exports[`update-interaction-metadata Everything is successful Walkthrough passes in the correct arguments to secretsClient.getSecretValue() to get cx credentials 1`] = `
Array [
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-cx",
    },
  ],
]
`;

exports[`update-interaction-metadata throws an error when there is a problem retrieving cx credentials 1`] = `"Error retrieving cx credentials"`;

exports[`update-interaction-metadata throws an error when there is a problem updating interaction metadata 1`] = `"Failed to update interaction metadata"`;
